File: TexasProgram/Services/ProTexasCalculator.swift
import Foundation

/// Продвинутый техасский метод Лимара: десять двухнедельных блоков.
///
/// Всё считается от веса интенсивного дня второй недели блока. Первые два блока
/// идут ниже твоего 5ПМ — это разгон, а не занижение: программу начинают после
/// нескольких месяцев классического техаса, и вход делается заведомо посильным.
enum ProTexasCalculator {
    static let blockCount = 10
    static let weekCount = blockCount * 2

    /// Сдвиг веса интенсивного дня по блокам: −20, −10, свой 5ПМ, дальше по 2,5 кг.
    static let blockOffsets: [Double] = [-20, -10, 0, 2.5, 5, 7.5, 10, 12.5, 15, 17.5]

    private static let memo = PlanMemo<ProgramInput, WorkoutPlan>()

    static func generate(input: ProgramInput) -> WorkoutPlan {
        memo.resolve(input) { value in
            WorkoutPlan(weeks: (1...weekCount).map { week($0, input: value) }, isPeaking: false)
        }
    }

    /// Номер двухнедельного блока для календарной недели.
    static func block(forWeek week: Int) -> Int { (week + 1) / 2 }

    /// Вес интенсивного дня блока — точка отсчёта для всех остальных дней.
    static func intensityWeight(_ base: Double, block: Int) -> Double {
        let offset = blockOffsets.indices.contains(block - 1) ? blockOffsets[block - 1] : blockOffsets.last ?? 0
        return ProgramCalculator.roundToPlate(base + offset)
    }

    private static func week(_ number: Int, input: ProgramInput) -> WorkoutWeekPlan {
        let blockNumber = block(forWeek: number)
        let squat = intensityWeight(input.squat5RM, block: blockNumber)
        let bench = intensityWeight(input.bench5RM, block: blockNumber)
        let deadlift = intensityWeight(input.deadlift5RM, block: blockNumber)
        // Нечётная неделя — первая в блоке, подготовительная; чётная — классическая.
        let isFirstWeek = number % 2 == 1

        let days = isFirstWeek
            ? firstWeek(input: input, squat: squat, bench: bench, deadlift: deadlift)
            : secondWeek(input: input, squat: squat, bench: bench, deadlift: deadlift)
        return WorkoutWeekPlan(id: number, number: number, days: days.enumerated().map { index, exercises in
            day(number, index + 1, dayTitle(isFirstWeek: isFirstWeek, dayNumber: index + 1), exercises)
        })
    }

    private static func dayTitle(isFirstWeek: Bool, dayNumber: Int) -> String {
        let phase = isFirstWeek ? "ПОДГОТОВКА" : "КЛАССИКА"
        switch dayNumber {
        case 1: return phase + " · ОБЪЁМ"
        case 2: return phase + " · ВОССТАНОВЛЕНИЕ"
        default: return phase + " · ИНТЕНСИВНОСТЬ"
        }
    }

    // MARK: - Первая неделя блока: субмаксимальная

    private static func firstWeek(input: ProgramInput, squat: Double, bench: Double, deadlift: Double) -> [[ExercisePrescription]] {
        var volume = [
            main("Приседания", sets: 5, reps: "7", weight: squat, percent: 0.8),
            main("Жим лёжа", sets: 5, reps: "7", weight: bench, percent: 0.8)
        ]
        append(&volume, optional(input.pull, sets: 4, reps: "7–10", rpe: "RPE 7"))
        append(&volume, optional(input.core, sets: 4, reps: "5–8", rpe: "RPE 8"))

        var recovery = [
            main("Приседания", sets: 2, reps: "5", weight: squat, percent: 0.7),
            main("Жим лёжа", sets: 2, reps: "5", weight: bench, percent: 0.7)
        ]
        append(&recovery, optional(input.back, sets: 3, reps: "8–12", rpe: "RPE 8"))
        append(&recovery, optional(input.press, sets: 3, reps: "7–10", rpe: "RPE 7"))
        append(&recovery, optional(input.arms, sets: 4, reps: "9–12", rpe: "RPE 8"))

        var intensity = [
            main("Приседания", sets: 3, reps: "3", weight: squat, percent: 0.9),
            main("Жим лёжа", sets: 3, reps: "3", weight: bench, percent: 0.9),
            main("Становая тяга", sets: 3, reps: "3", weight: deadlift, percent: 0.9)
        ]
        append(&intensity, optional(input.back, sets: 3, reps: "8–12", rpe: "RPE 8"))
        append(&intensity, optional(input.core, sets: 4, reps: "13–16", rpe: "RPE 8"))

        return [volume, recovery, intensity]
    }

    // MARK: - Вторая неделя блока: классический техас

    private static func secondWeek(input: ProgramInput, squat: Double, bench: Double, deadlift: Double) -> [[ExercisePrescription]] {
        var volume = [
            main("Приседания", sets: 5, reps: "5", weight: squat, percent: 0.9),
            main("Жим лёжа", sets: 5, reps: "5", weight: bench, percent: 0.9)
        ]
        append(&volume, optional(input.pull, sets: 2, reps: "5", rpe: "RPE 7"))
        append(&volume, optional(input.core, sets: 3, reps: "5–8", rpe: "RPE 8"))

        var recovery = [
            main("Приседания", sets: 2, reps: "5", weight: squat, percent: 0.7),
            main("Жим лёжа", sets: 2, reps: "5", weight: bench, percent: 0.7)
        ]
        append(&recovery, optional(input.back, sets: 3, reps: "5–8", rpe: "RPE 8"))
        append(&recovery, optional(input.press, sets: 3, reps: "5", rpe: "RPE 7"))
        append(&recovery, optional(input.arms, sets: 3, reps: "9–12", rpe: "RPE 8"))

        // Кульминация блока: истинный пятиповторный максимум.
        var intensity = [
            main("Приседания", sets: 1, reps: "5", weight: squat, percent: 1),
            main("Жим лёжа", sets: 1, reps: "5", weight: bench, percent: 1),
            main("Становая тяга", sets: 1, reps: "5", weight: deadlift, percent: 1)
        ]
        append(&intensity, optional(input.back, sets: 3, reps: "8–12", rpe: "RPE 8"))
        append(&intensity, optional(input.core, sets: 3, reps: "13–16", rpe: "RPE 8"))

        return [volume, recovery, intensity]
    }

    // MARK: - Сборка

    private static func main(_ name: String, sets: Int, reps: String, weight: Double, percent: Double) -> ExercisePrescription {
        ExercisePrescription(
            name: name,
            sets: sets,
            reps: reps,
            load: .kilograms(ProgramCalculator.roundToPlate(weight * percent))
        )
    }

    private static func optional(_ name: String?, sets: Int, reps: String, rpe: String) -> ExercisePrescription? {
        guard let name, !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
        return ExercisePrescription(name: name, sets: sets, reps: reps, load: .rpe(rpe), isOptional: true)
    }

    private static func append(_ array: inout [ExercisePrescription], _ value: ExercisePrescription?) {
        if let value { array.append(value) }
    }

    private static func day(_ week: Int, _ number: Int, _ title: String, _ exercises: [ExercisePrescription]) -> WorkoutDayPlan {
        WorkoutDayPlan(id: "\(week)-\(number)", number: number, title: title, exercises: exercises)
    }
}

File: android/app/src/main/java/com/texasprogram/app/service/ProTexasCalculator.kt
package com.texasprogram.app.service

import com.texasprogram.app.model.ExercisePrescription
import com.texasprogram.app.model.LoadPrescription
import com.texasprogram.app.model.ProgramInput
import com.texasprogram.app.model.WorkoutDayPlan
import com.texasprogram.app.model.WorkoutPlan
import com.texasprogram.app.model.WorkoutWeekPlan

/// Продвинутый техасский метод Лимара: десять двухнедельных блоков.
///
/// Всё считается от веса интенсивного дня второй недели блока. Первые два блока
/// идут ниже твоего 5ПМ — это разгон, а не занижение: программу начинают после
/// нескольких месяцев классического техаса, и вход делается заведомо посильным.
object ProTexasCalculator {
    const val BLOCK_COUNT = 10
    const val WEEK_COUNT = BLOCK_COUNT * 2

    /// Сдвиг веса интенсивного дня по блокам: −20, −10, свой 5ПМ, дальше по 2,5 кг.
    val blockOffsets = listOf(-20.0, -10.0, 0.0, 2.5, 5.0, 7.5, 10.0, 12.5, 15.0, 17.5)

    private var cacheKey: Any? = null
    private var cacheValue: WorkoutPlan? = null

    fun generate(input: ProgramInput): WorkoutPlan {
        cacheValue?.let { if (cacheKey == input) return it }
        val plan = WorkoutPlan(weeks = (1..WEEK_COUNT).map { week(it, input) }, isPeaking = false)
        cacheKey = input
        cacheValue = plan
        return plan
    }

    /// Номер двухнедельного блока для календарной недели.
    fun block(week: Int): Int = (week + 1) / 2

    /// Вес интенсивного дня блока — точка отсчёта для всех остальных дней.
    fun intensityWeight(base: Double, block: Int): Double {
        val offset = blockOffsets.getOrElse(block - 1) { blockOffsets.last() }
        return ProgramCalculator.roundToPlate(base + offset)
    }

    private fun week(number: Int, input: ProgramInput): WorkoutWeekPlan {
        val blockNumber = block(number)
        val squat = intensityWeight(input.squat5RM, blockNumber)
        val bench = intensityWeight(input.bench5RM, blockNumber)
        val deadlift = intensityWeight(input.deadlift5RM, blockNumber)
        // Нечётная неделя — первая в блоке, подготовительная; чётная — классическая.
        val isFirstWeek = number % 2 == 1

        val days = if (isFirstWeek) firstWeek(input, squat, bench, deadlift)
        else secondWeek(input, squat, bench, deadlift)

        return WorkoutWeekPlan(
            number = number,
            days = days.mapIndexed { index, exercises ->
                WorkoutDayPlan("$number-${index + 1}", index + 1, dayTitle(isFirstWeek, index + 1), exercises)
            }
        )
    }

    private fun dayTitle(isFirstWeek: Boolean, dayNumber: Int): String {
        val phase = if (isFirstWeek) "ПОДГОТОВКА" else "КЛАССИКА"
        return phase + when (dayNumber) {
            1 -> " · ОБЪЁМ"
            2 -> " · ВОССТАНОВЛЕНИЕ"
            else -> " · ИНТЕНСИВНОСТЬ"
        }
    }

    // MARK: - Первая неделя блока: субмаксимальная

    private fun firstWeek(input: ProgramInput, squat: Double, bench: Double, deadlift: Double): List<List<ExercisePrescription>> {
        val volume = mutableListOf(
            main("Приседания", 5, "7", squat, 0.8),
            main("Жим лёжа", 5, "7", bench, 0.8)
        )
        add(volume, optional(input.pull, 4, "7–10", "RPE 7"))
        add(volume, optional(input.core, 4, "5–8", "RPE 8"))

        val recovery = mutableListOf(
            main("Приседания", 2, "5", squat, 0.7),
            main("Жим лёжа", 2, "5", bench, 0.7)
        )
        add(recovery, optional(input.back, 3, "8–12", "RPE 8"))
        add(recovery, optional(input.press, 3, "7–10", "RPE 7"))
        add(recovery, optional(input.arms, 4, "9–12", "RPE 8"))

        val intensity = mutableListOf(
            main("Приседания", 3, "3", squat, 0.9),
            main("Жим лёжа", 3, "3", bench, 0.9),
            main("Становая тяга", 3, "3", deadlift, 0.9)
        )
        add(intensity, optional(input.back, 3, "8–12", "RPE 8"))
        add(intensity, optional(input.core, 4, "13–16", "RPE 8"))

        return listOf(volume, recovery, intensity)
    }

    // MARK: - Вторая неделя блока: классический техас

    private fun secondWeek(input: ProgramInput, squat: Double, bench: Double, deadlift: Double): List<List<ExercisePrescription>> {
        val volume = mutableListOf(
            main("Приседания", 5, "5", squat, 0.9),
            main("Жим лёжа", 5, "5", bench, 0.9)
        )
        add(volume, optional(input.pull, 2, "5", "RPE 7"))
        add(volume, optional(input.core, 3, "5–8", "RPE 8"))

        val recovery = mutableListOf(
            main("Приседания", 2, "5", squat, 0.7),
            main("Жим лёжа", 2, "5", bench, 0.7)
        )
        add(recovery, optional(input.back, 3, "5–8", "RPE 8"))
        add(recovery, optional(input.press, 3, "5", "RPE 7"))
        add(recovery, optional(input.arms, 3, "9–12", "RPE 8"))

        // Кульминация блока: истинный пятиповторный максимум.
        val intensity = mutableListOf(
            main("Приседания", 1, "5", squat, 1.0),
            main("Жим лёжа", 1, "5", bench, 1.0),
            main("Становая тяга", 1, "5", deadlift, 1.0)
        )
        add(intensity, optional(input.back, 3, "8–12", "RPE 8"))
        add(intensity, optional(input.core, 3, "13–16", "RPE 8"))

        return listOf(volume, recovery, intensity)
    }

    private fun main(name: String, sets: Int, reps: String, weight: Double, percent: Double) =
        ExercisePrescription(name, sets, reps, LoadPrescription.Kilograms(ProgramCalculator.roundToPlate(weight * percent)))

    private fun optional(name: String?, sets: Int, reps: String, rpe: String): ExercisePrescription? {
        if (name.isNullOrBlank()) return null
        return ExercisePrescription(name, sets, reps, LoadPrescription.Rpe(rpe), isOptional = true)
    }

    private fun add(list: MutableList<ExercisePrescription>, value: ExercisePrescription?) {
        if (value != null) list.add(value)
    }
}
